File: base/conditions.py
from dfexcept import DragonflyException
from PyQt5.QtXml import QDomDocument, QDomElement
import action

class IsSet(action.Condition):
	def __init__(self) -> None:
		super().__init__()

		self.instance = ""
		self.attr = ""

	def __str__(self) -> str:
		return f'Is set "{self.attr}" on "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		# Gets the noun
		noun = action.dictionary.nouns(self.instance)
		if not noun:
			raise DragonflyException(f'On IsSet condition: instance "{self.instance}" not found in dictionary.')
	
		return noun[0].isSet(self.attr)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance")
		self.attr = element.attribute("attr")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("instance", self.instance)
		element.setAttribute("attr", self.attr)
		return element

class DirectEqualsExit(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.exit = ""

	def __str__(self) -> str:
		return f'Direct equals exit: "{self.exit}"'

	def check(self, action: "action.Action") -> bool:
		e = action.dictionary.exit(self.exit)
		if not e:
			raise DragonflyException(f'On DirectEqualsExit condition: exit "{self.exit}" not found in dictionary.')

		return e.responds(action.parser.directObjectString)

	def load(self, element: QDomElement):
		self.exit = element.attribute("exit")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("exit", self.exit)
		return element
		
class Contains(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.container = ""
		self.instance = ""

	def __str__(self) -> str:
		return f'"{self.container}" contains "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		cont = action.dictionary.nouns(self.container)
		if not cont:
			raise DragonflyException(f'On Contains condition: container "{self.container}" not found in dictionary.')

		return cont[0].contains(self.instance)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance", defaultValue="")
		self.container = element.attribute("container", defaultValue="")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)

		element.setAttribute("instance", self.instance)
		element.setAttribute("container", self.container)

		return element

class DirectEquals(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.instance = ""

	def __str__(self) -> str:
		return f'Direct equals "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		obj = action.parser.directObject
		if not obj: return False

		return obj.responds(self.instance)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance", defaultValue="")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("instance", self.instance)
		return element

class DirectNotEquals(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.instance = ""

	def __str__(self) -> str:
		return f'Direct not equals "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		obj = action.parser.directObject
		if not obj: return True

		return not obj.responds(self.instance)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance", defaultValue="")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("instance", self.instance)
		return element

class IndirectEquals(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.instance = ""

	def __str__(self) -> str:
		return f'Indirect equals "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		obj = action.parser.indirectObject
		if not obj: return False

		return obj.responds(self.instance)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance", defaultValue="")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("instance", self.instance)
		return element

class IndirectNotEquals(action.Condition):
	def __init__(self) -> None:
		super().__init__()
		self.instance = ""

	def __str__(self) -> str:
		return f'Indirect equals "{self.instance}"'

	def check(self, action: "action.Action") -> bool:
		obj = action.parser.indirectObject
		if not obj: return True

		return not obj.responds(self.instance)

	def load(self, element: QDomElement):
		self.instance = element.attribute("instance", defaultValue="")

	def save(self, doc: QDomDocument) -> QDomElement:
		element = super().save(doc)
		element.setAttribute("instance", self.instance)
		return element

File: base/conditions.js
import { Condition } from "./condition.js";
import { Output } from "./output.js";

export const conditions = {};

export class IsSet extends Condition {
	constructor() {
		super();

		this.instance = "";
		this.attr = "";
	}

	toString() {
		return `Is set "${this.attr}" on "${this.instance}"`;
	}

	check(action) {
		// Gets the noun
		const noun = action.book.dictionary.getNouns(this.instance);
		if (noun.length === 0) {
			 
			Output.error(`On IsSet condition: instance "${this.instance}" not found in dictionary.`);
		}
	
		return noun[0].isSet(this.attr);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.attr = node.getAttr("attr").getValue();
	}
} conditions.IsSet = IsSet;

export class IsNotSet extends Condition {
	constructor() {
		super();

		this.instance = "";
		this.attr = "";
	}

	toString() {
		return `Is not set "${this.attr}" on "${this.instance}"`;
	}

	check(action) {
		// Gets the noun
		const noun = action.book.dictionary.getNouns(this.instance)
		if (noun.length === 0) {
			 
			Output.error(`On IsNotSet condition: instance "${this.instance}" not found in dictionary.`);
		}
		return !noun[0].isSet(this.attr);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.attr = node.getAttr("attr").getValue();
	}
} conditions.IsNotSet = IsNotSet;

export class DirectEqualsExit extends Condition {
	constructor() {
		super();
		this.exit = "";
	}

	toString() {
		return `Direct equals exit: "${this.exit}"`;
	}

	check(action) {
		const e = action.book.dictionary.getExit(this.exit);
		if (!e) {
			 
			Output.error(`On DirectEqualsExit condition: exit "${this.exit}" not found in dictionary.`);
		}

		return e.responds(action.book.parser.directObjectString);
	}

	load(node) {
		this.exit = node.getAttr("exit").getValue();
	}
} conditions.DirectEqualsExit = DirectEqualsExit;
		
export class Contains extends Condition {
	constructor() {
		super();
		this.container = "";
		this.instance = "";
	}

	toString() {
		return `"${this.container}" contains "${this.instance}"`;
	}

	check(action) {
		cont = action.book.dictionary.getNouns(this.container)
		if (cont.length === 0) {
			 
			Output.error(`On Contains condition: container "${this.container}" not found in dictionary.`);
		}

		return cont[0].contains(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.container = node.getAttr("container").getValue();
	}
} conditions.Contains = Contains;

export class NotContains extends Condition {
	constructor() {
		super();
		this.container = "";
		this.instance = "";
	}

	toString() {
		return `"${this.container}" not contains "${this.instance}"`;
	}

	check(action) {
		cont = action.book.dictionary.getNouns(this.container)
		if (cont.length === 0) {
			
			Output.error(`On NotContains condition: container "${this.container}" not found in dictionary.`);
		}

		return !cont[0].contains(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.container = node.getAttr("container").getValue();
	}
} conditions.NotContains = NotContains;

export class DirectEquals extends Condition {
	constructor() {
		super();
		this.instance = "";
	}

	toString() {
		return `Direct equals "${this.instance}"`;
	}

	check(action) {
		const obj = action.book.parser.directObject;
		if (!obj) return false;

		return obj.responds(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
	}
} conditions.DirectEquals = DirectEquals;

export class DirectNotEquals extends Condition {
	constructor() {
		super();
		this.instance = "";
	}

	toString() {
		return `Direct not equals "${this.instance}"`;
	}

	check(action) {
		obj = action.book.parser.directObject;
		if (!obj) return true;

		return !obj.responds(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
	}
} conditions.DirectNotEquals = DirectNotEquals;

export class IndirectEquals extends Condition {
	constructor() {
		super();
		this.instance = "";
	}

	toString() {
		return `Indirect equals "${this.instance}"`;
	}

	check(action) {
		const obj = action.book.parser.indirectObject;
		if (!obj) return false;

		return obj.responds(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
	}
} conditions.IndirectEquals = IndirectEquals;

export class IndirectNotEquals extends Condition {
	constructor() {
		super();
		this.instance = "";
	}

	toString() {
		return `Indirect equals "${this.instance}"`;
	}

	check(action) {
		const obj = action.book.parser.indirectObject;
		if (!obj) return true;

		return !obj.responds(this.instance);
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
	}
} conditions.IndirectNotEquals = IndirectNotEquals;

export class VariableEquals extends Condition {
	constructor() {
		super();
		this.instance = "";
		this.variable = "";
		this.value = "";
	}

	toString() {
		return `Variable "${this.variable}" equals to "${this.value}."`;
	}

	check(action) {
		const obj = action.book.dictionary.getNouns(this.instance);

		if (obj.length === 0) {
			
			Output.error(`On condition "VariableEquals" instance "${this.instance}" not found in dictionary.`);
		}

		return obj[0].getVariable(this.variable) === this.value;
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.variable = node.getAttr("variable").getValue();
		this.value = node.getAttr("value").getValue();
	}
} conditions.VariableEquals = VariableEquals;

export class ConnectionExists extends Condition {
	constructor() {
		super();
		this.instance = "";
		this.exit = "";
	}

	toString() {
		return `Connection "${this.exit}" exists in "${this.instance}."`;
	}

	check(action) {
		const place = action.book.dictionary.getNouns(this.instance);
		if (place.length === 0) {
			Output.error(`On condition "ConnectionExists" instance "${this.instance}" not found in dictionary.`);
		}

		return place[0].getConnection(this.exit) != null;
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.exit = node.getAttr("exit").getValue();
	}

} conditions.ConnectionExists = ConnectionExists;

export class ConnectionNotExists extends Condition {
	constructor() {
		super();
		this.instance = "";
		this.exit = "";
	}

	toString() {
		return `Connection "${this.exit}" not exists in "${this.instance}."`;
	}

	check(action) {
		const place = action.book.dictionary.getNouns(this.instance);
		if (place.length === 0) {
			Output.error(`On condition "ConnectionNotExists" instance "${this.instance}" not found in dictionary.`);
		}

		return place[0].getConnection(this.exit) == null;
	}

	load(node) {
		this.instance = node.getAttr("instance").getValue();
		this.exit = node.getAttr("exit").getValue();
	}

} conditions.ConnectionNotExists = ConnectionNotExists;
